loupe: init at unstable-2023-03-14

06kellyjac opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Package loupe for nixpkgs

Things to think about:

* try build for other platforms and arches?
* wait for a tag?
* does it also want dbus? pretty much everything seemed to work but there was an issue with the GUI "help" feature

>  2023-03-15T12:39:07.066Z ERROR loupe::application > Failed to launch help: The specified location is not supported

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

06kellyjac commented:
Result of `nixpkgs-review pr 202692` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>loupe</li>
  </ul>
</details>

BenediktBroich commented:
Result of `nixpkgs-review pr 202692` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>loupe</li>
  </ul>
</details>

jtojnar commented on pkgs/applications/graphics/loupe/default.nix:
```diff
@@ -0,0 +1,61 @@
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "loupe";
+  version = "44.alpha";
```
According to https://nixos.org/nixpkgs/manual/#sec-package-naming, the `version` should be `unstable-yyyy-mm-dd`, when using untagged git snapshots.

06kellyjac replied:
I usually go with that but loupe has it's own version listed so I went with that but I can change it

jtojnar commented on pkgs/applications/graphics/loupe/default.nix:
```diff
@@ -0,0 +1,61 @@
+  pname = "loupe";
+  version = "44.alpha";
+
+  src = fetchgit {
```
Use `fetchGitLab` with appropriate `domain`.

jtojnar commented on pkgs/applications/graphics/loupe/default.nix:
```diff
@@ -0,0 +1,61 @@
+    itstool
+    desktop-file-utils # update-desktop-database
+    ninja
+    gnome-builder
```
This sounds extremely suspicious. Builder is GNOME’s IDE.

06kellyjac replied:
yeah I just saw it on the README as part of development & I'm not very familiar with gtk software builds

removed

jtojnar commented on pkgs/applications/graphics/loupe/default.nix:
```diff
@@ -0,0 +1,61 @@
+  nativeBuildInputs = [
+    meson
+    pkg-config
+    cmake
```
Are you sure `cmake` is needed? Meson supports it for finding dependencies but typically, `pkg-config` will be sufficient. See <https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/meson-cmake.md> for more info.

06kellyjac replied:
hmmm, it 100% asked for `cmake` at one point but yeah now it's happy without. I guess adding `pkg-config` or `ninja` made it stop

jtojnar commented on pkgs/applications/graphics/loupe/default.nix:
```diff
@@ -0,0 +1,61 @@
+    description = "A simple image viewer application written with GTK4 and Rust";
+    license = licenses.gpl3Plus;
+    maintainers = with maintainers; [ jk ];
+    platforms = [ "x86_64-linux" ];
```
Does it not work on other platforms?

06kellyjac replied:
Probably does but there's not much docs and that's the only platform I've tried so far

BenediktBroich replied:
Most of the other GTK3/4 applications in nixpkgs have platforms.unix here.

jtojnar commented on pkgs/applications/graphics/loupe/default.nix:
```diff
@@ -0,0 +1,61 @@
+    rust.rustc
+  ]);
+
+  buildInputs = [
```
Not sure how Loupe loads images but if it relies on `gdk-pixbuf` internally, you might want to look on `eog` for how to deal with `webp` support.

06kellyjac replied:
It does use gdk-pixbuf but it doesn't seem to support gif and webp yet so I think I can save that for another time

https://gitlab.gnome.org/BrainBlasted/loupe/-/issues/27

---

My main outstanding question is about `./result/share/dbus-1/services/org.gnome.Loupe.service` & `loupe --gapplication-service`

jtojnar replied:
That sounds like the issue is just about animation. Most webp images should be just static: https://developers.google.com/speed/webp/gallery1

06kellyjac replied:
`ERROR loupe::widgets::image_page > Could not load image: Unrecognised image file format`

I guess I could try the `eog` bits and see if it fixes things

06kellyjac replied:
I've added the eog parts and I was able to open a webp file

IDK if I needed all of these parts so I've put it in a separate WIP commit.
I might look at trimming it down depending on what is actually necessary

06kellyjac committed:
loupe: init at unstable-2023-03-01

BenediktBroich commented:
Result of `nixpkgs-review pr 202692` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>loupe</li>
  </ul>
</details>

06kellyjac commented:
Still need to make some fixes

06kellyjac committed:
WIP add webp support

jtojnar committed:
gtk4: 4.8.3 → 4.9.4

https://gitlab.gnome.org/GNOME/gtk/-/compare/4.8.3...4.9.4

Changelog-Reviewed-By: Jan Tojnar <jtojnar@gmail.com>

jtojnar committed:
libadwaita: 1.2.3 → 1.3.beta

https://gitlab.gnome.org/GNOME/libadwaita/-/compare/1.2.3...1.3.beta

Changelog-Reviewed-By: Jan Tojnar <jtojnar@gmail.com>

06kellyjac committed:
loupe: unstable-2023-03-01 -> unstable-2023-03-14

06kellyjac commented:
libgweather is now required "to determine the city for EXIF GPS coordinates."

https://gitlab.gnome.org/Incubator/loupe/-/commit/65808f58945b7552537c75dc586e0f9427183d90

---

I haven't tried minifying the [WIP add webp support](https://github.com/NixOS/nixpkgs/pull/202692/commits/89f4b1c3a6030bf721f4a14e45d7f386e0d732c2) changes and at this point I might not bother, just keep what eog has :shrug:

---

latest maeson.build requires gtk4 >=4.9.4
and gtk4-sys with feature v4_10 requires gtk4 >= 4.9

I tried updating to an older release 2023-03-01 which built fine but crashes when trying to run it.
I decided to pull in changes from #218143 and just build the latest which worked fine

---

Going to keep this as draft certainly until #218143 makes it to master and maybe until loupe tags off a release? I'm not sure how the incubation process goes

Link might be useful: https://gitlab.gnome.org/Incubator/Submission/-/issues/1

There are some milestones/project boards:

[Gnome GitLab - Milestones - Core Target](https://gitlab.gnome.org/Incubator/loupe/-/milestones/3#tab-issues)
[Gnome GitLab - Milestones - Core Blocker](https://gitlab.gnome.org/Incubator/loupe/-/milestones/2#tab-issues)

This also might be useful:

![image](https://user-images.githubusercontent.com/9866621/225310543-5bb49fdc-df70-4182-96db-ce7cc10e7329.png)

And some things we might want to test:

[#61 - Flatpak Limitations](https://gitlab.gnome.org/Incubator/loupe/-/issues/61)

---

cc @jtojnar

Would you consider the "Incubation" category as under the gnome maintainer group or would you rather have a separate group or no group and wait until these projects graduate?

jtojnar commented:
I think waiting for graduation is fine.

By the way, the gdk-pixbuf thing should no longer be needed according to https://thisweek.gnome.org/posts/2023/03/twig-86/#loupe

06kellyjac commented:
that's good to know. Should I keep (most) of the thumbnailer parts though?

jtojnar commented:
I would not expect the thumbnailers to be used for anything. I guess the only place that could use them is the file chooser but that is used through portal so it will not see thumbnailers defined in the app.

chuangzhu commented:
There is an initial release now: https://gitlab.gnome.org/Incubator/loupe/-/releases/44.0

06kellyjac commented:
Thanks for making me aware.
Ill give it an update but they typically use latest rust so might need to wait for rust 1.68 to reach unstable

chuangzhu commented on pkgs/applications/graphics/loupe/default.nix:
```diff
@@ -0,0 +1,94 @@
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "loupe";
+  version = "unstable-2023-03-14";
```
```suggestion
  version = "44.0";
```

chuangzhu commented on pkgs/applications/graphics/loupe/default.nix:
```diff
@@ -0,0 +1,94 @@
+    domain = "gitlab.gnome.org";
+    owner = "Incubator";
+    repo = "loupe";
+    rev = "889c818c14179e3f4755fb56f38d1af603e009cf";
+    hash = "sha256-9FNUAypm2GAd8/deEbNhCV+eIEGxOBCzFyytKaK9dnw=";
```
```suggestion
    rev = finalAttrs.version;
    hash = "sha256-0jGffLjewybXljz58R9wuOKKOq56nzlxztgYh0fYrzM=";
```

chuangzhu commented on pkgs/applications/graphics/loupe/default.nix:
```diff
@@ -0,0 +1,94 @@
+    hash = "sha256-9FNUAypm2GAd8/deEbNhCV+eIEGxOBCzFyytKaK9dnw=";
+  };
+
+  cargoDeps = rustPlatform.fetchCargoTarball {
+    inherit (finalAttrs) src;
+    hash = "sha256-tUi+y6raAhfG1bvUuSR/MRIax3OhExOBxn8gElLBRfI=";
+  };
```
```suggestion
  cargoDeps = rustPlatform.importCargoLock {
    lockFile = finalAttrs.src + /Cargo.lock;
    outputHashes = {
      "librsvg-2.56.0" = "sha256-4poP7xsoylmnKaUWuJ0tnlgEMpw9iJrM3dvt4IaFi7w=";
    };
  };
```

jtojnar replied:
> `lockFile = finalAttrs.src + /Cargo.lock;`

This will use import-from-derivation, which is not allowed in Nixpkgs.

06kellyjac replied:
Yeah need to vendor the cargo.lock in this case

chuangzhu reviewed (commented):
This builds for me

bobby285271 commented:
I pushed 3412224d460822f9f5213f9abecc1b57e7b3432d + f52609aaaaa96b75c3a05e5bb0a1a6fbe716c7ed + 3b0feaa94f97e64f3eb2d6c5ba386e184c21a2d4 on gnome branch for now.
